src/NuGet.Services.Staging.Search/DatabaseSearchExpressionBuilder.cs

```diff
@@ -0,0 +1,143 @@
+                { QueryField.Version, VersionCriteria },
+                { QueryField.Title, TitleCriteria },
+                { QueryField.Description, DescriptionCriteria },
+                { QueryField.Tag, TagCriteria },
+                { QueryField.Author, AuthorCriteria },
+                { QueryField.Summary, SummaryCriteria },
+                { QueryField.Owner, OwnerCriteria },
+            };
+
+        public static Expression<Func<PackageMetadata, bool>> Build(string query)
+        {
+            // Parse the query string, so that a string like "TITLE:dot net ID:json" becomes
+            // Grouping 1: Key=TITLE VALUE=[dot, net]
+            // Grouping 2: Key=ID VALUE=[json] 
+            var parser = new NuGetQueryParser();
+            var groupings = parser.ParseQuery(query);
+
+            // Empty query. Return everything
+            if (!groupings.Any())
+            {
+                return p => true;
+            }
+
+            // Build a list of expressions for each term
+            var expressions = new List<LambdaExpression>();
+
+            foreach (var grouping in groupings)
+            {
+                Criteria fieldCriteria;
+                if (QueryFieldToCriteriaMapping.TryGetValue(grouping.Key, out fieldCriteria))
+                {
+                    foreach (var value in grouping.Value)
+                    {
+                        string formatedValue;
+                        if (!TryFormatValueForQuery(grouping.Key, value, out formatedValue))
+                        {
+                            // Illegal query. Result should be empty
+                            return p => false;
+                        }
+                        expressions.Add(fieldCriteria(formatedValue));
+                    }
+                }
+                else
+                {
+                    foreach (var criteria in QueryFieldToCriteriaMapping.Values)
+                    {
+                        expressions.AddRange(grouping.Value.Select(value => criteria(value)));
+                    }
+                }
+            }
+
+            // Build a giant or statement using the bodies of the lambdas
+            var body = expressions.Select(p => p.Body).Aggregate(Expression.OrElse);
+
+            // Now build the final predicate
+            var parameterExpr = Expression.Parameter(typeof(PackageMetadata));
+
+            // Fix up the body to use our parameter expression
+            body = new ParameterExpressionReplacer(parameterExpr).Visit(body);
+
+            // Build the final predicate
+            var predicate = Expression.Lambda<Func<PackageMetadata, bool>>(body, parameterExpr);
+
+            return predicate;
+        }
+
+        private static bool TryFormatValueForQuery(QueryField field, string value, out string formatedValue)
+        {
+            formatedValue = value;
+
+            if (field == QueryField.Version)

```

Comment:
From the code I see here: https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Services/Extensions.cs
there is no support in the gallery for searching by version...